synapse/replication/tcp/handler.py

```diff
@@ -314,17 +315,19 @@ def start_replication(self, hs):
             self._factory = RedisDirectTcpReplicationClientFactory(
                 hs, outbound_redis_connection
             )
-            hs.get_reactor().connectTCP(
+            endpoint = HostnameEndpoint(
+                hs.get_reactor(),
                 hs.config.redis.redis_host.encode(),
                 hs.config.redis.redis_port,
-                self._factory,
             )
+            endpoint.connect(self._factory)
         else:
             client_name = hs.get_instance_name()
             self._factory = DirectTcpReplicationClientFactory(hs, client_name, self)
             host = hs.config.worker_replication_host
             port = hs.config.worker_replication_port
-            hs.get_reactor().connectTCP(host.encode(), port, self._factory)
+            endpoint = HostnameEndpoint(hs.get_reactor(), host.encode(), port)
+            endpoint.connect(self._factory)
```

Comment:
Unfortunately, I think fixing this is more involved. It's to do with how we handle connection failures, and retries.

`DirectTcpReplicationClientFactory` and `RedisDirectTcpReplicationClientFactory` are both based on `twisted.internet.protocol.ReconnectingClientFactory`. If `reactor.connectTCP` is unable to connect to the hostname, it will call `ReconnectingClientFactory.clientConnectionFailed`, which will schedule a retry a few seconds later.

`HostnameEndpoint.connect` does its own set of retries - one for each IP address returned by the hostname lookup - and then calls `reactor.connectTCP` for each.

So, if we use `HostnameEndpoint` *and* `ReconnectingClientFactory`, we can end up with two different, conflicting, retry systems going on at the same time, leaving us in a terrible mess.

I gather (h/t @glyph) that `ReconnectingClientFactory` is softly deprecated, and instead we want to use https://twistedmatrix.com/documents/current/api/twisted.application.internet.ClientService.html, which will compose properly with a `HostnameEndpoint`.

So, I think what we want to do is create a `ClientService` in the constructor of `ReplicationCommandHandler`, and then we can just call `ClientService.startService` in `start_replication`.

---

Now, all of this is getting rather complicated. I think you should split up this PR - let's just fix the SMTP sender for now, and then consider fixing the replication connections in a later PR?